pkg/externalpipeline/externalpipeline.go

```diff
@@ -0,0 +1,167 @@
+package externalpipeline
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/RedHatInsights/insights-results-smart-proxy/types"
+	"k8s.io/client-go/pkg/version"
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	"github.com/openshift/insights-operator/pkg/config"
+)
+
+type ReportsCache struct {
+	Configuration config.SmartProxy
+	client        *http.Client
+	authorizer    Authorizer
+	clusterInfo   ClusterVersionInfo
+	LastReport    types.SmartProxyReport
+}
+
+type InsightsReportResponse struct {
+	Report types.SmartProxyReport `json:"report"`
+	Status string                 `json:"status"`
+}
+
+type ClusterVersionInfo interface {
+	ClusterVersion() *configv1.ClusterVersion
+}
+
+type Authorizer interface {
+	Authorize(req *http.Request) error
+	NewSystemOrConfiguredProxy() func(*http.Request) (*url.URL, error)
+}
+
+var (
+	// ErrWaitingForVersion contains an error when the cluster version is not loaded yet
+	ErrWaitingForVersion = fmt.Errorf("waiting for the cluster version to be loaded")
+
+	// insightsStatus contains a metric with the latest report information
+	insightsStatus = metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Namespace: "health",
+		Subsystem: "statuses",
+		Name:      "insights",
+		Help:      "Information about the cluster health status as detected by Insights tooling.",
+	}, []string{"metric"})
+)
+
+func New(c config.SmartProxy, authorizer Authorizer, clusterInfo ClusterVersionInfo) *ReportsCache {
+	return &ReportsCache{
+		Configuration: c,
+		client:        &http.Client{},
+		authorizer:    authorizer,
+		clusterInfo:   clusterInfo,
+	}
+}
+
+func (r *ReportsCache) PullSmartProxy() error {
+	klog.Info("Pulling report from smart-proxy")
+	cv := r.clusterInfo.ClusterVersion()
+	if cv == nil {
+		klog.Warning(ErrWaitingForVersion)
+		return ErrWaitingForVersion
+	}
+
+	endpoint := fmt.Sprintf(r.Configuration.Endpoint, cv.Spec.ClusterID)
+	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
+
+	klog.Infof("Retrieving report for cluster: %s", cv.Spec.ClusterID)
+	klog.Infof("Endpoint: %s", endpoint)
+	if err != nil {
+		klog.Error(err)
+		return err
+	}
+
+	if req.Header == nil {
+		req.Header = make(http.Header)
+	}
+	req.Header.Set("User-Agent", fmt.Sprintf("insights-operator/%s cluster/%s", version.Get().GitCommit, cv.Spec.ClusterID))
```

Comment:
Shouldn't the https://github.com/openshift/insights-operator/blob/master/pkg/insights/insightsclient/insightsclient.go be made a bit more generic to handle other requests as well? Seems like quite a lot of common code, and things like configuring a proxy is not even handled here: I suppose it's right expectation for the same transport mechanism for upload&pull the data: the only thing that differs is the endpoint/verb, doesn't it?

Reply:
I didn't study the source code of that package, but probably both are Insights related, so it will make sense to extract "common factor" and use this common code from both packages. Probably it worths to move the new `externalpipeline` package into the `insights` one, so some of the interface and common code should be easily reused.

I will take a look into it and try my best to see what can be done.

Right now I'm just fighting against the malevolous CI